worlds/yachtdice/Rules.py

```diff
@@ -0,0 +1,253 @@
+# This gives rise to an approximate discrete distribution per category.
+# We calculate the distribution of the total score.
+# We then pick a correct percentile to reflect the correct score that should be in logic.
+# The score is logic is *much* lower than the actual maximum reachable score.
+
+
+class Category:
+    def __init__(self, name, quantity=1):
+        self.name = name
+        self.quantity = quantity  # how many times you have the category
+
+    # return mean score of a category
+    def mean_score(self, num_dice, num_rolls):
+        if num_dice == 0 or num_rolls == 0:
+            return 0
+        mean_score = 0
+        for key in yacht_weights[self.name, min(8, num_dice), min(8, num_rolls)]:
+            mean_score += key * yacht_weights[self.name, min(8, num_dice), min(8, num_rolls)][key] / 100000
+        return mean_score * self.quantity
+
+
+class ListState:
+    def __init__(self, state: List[str]):
+        self.state = state
+
+    def count(self, item: str, player: Optional[str] = None) -> int:
+        return self.state.count(item)
+
+
+def extract_progression(state, player, options):
+    # method to obtain a list of what items the player has.
+    # this includes categories, dice, rolls and score multiplier etc.
+    # First, we convert the state if it's a list, so we can use state.count(item, player)
+    if isinstance(state, list):
+        state = ListState(state=state)
+
+    number_of_dice = (
+        state.count("Dice", player)
+        + state.count("Dice Fragment", player) // options.number_of_dice_fragments_per_dice.value
+    )
+    number_of_rerolls = (
+        state.count("Roll", player)
+        + state.count("Roll Fragment", player) // options.number_of_roll_fragments_per_roll.value
+    )
+    number_of_fixed_mults = state.count("Fixed Score Multiplier", player)
+    number_of_step_mults = state.count("Step Score Multiplier", player)
+    categories = []
+    for category_name, category_value in category_mappings.items():
+        if state.count(category_name, player) >= 1:
+            categories += [Category(category_value, state.count(category_name, player))]
+    extra_points_in_logic = state.count("1 Point", player)
+    extra_points_in_logic += state.count("10 Points", player) * 10
+    extra_points_in_logic += state.count("100 Points", player) * 100
+
+    return [
+        categories,
+        number_of_dice,
+        number_of_rerolls,
+        number_of_fixed_mults * 0.1,
+        number_of_step_mults * 0.01,
+        extra_points_in_logic,
+    ]
+
+
+# We will store the results of this function as it is called often for the same parameters.
+
+
+yachtdice_cache = {}
+
+
+# Function that returns the feasible score in logic based on items obtained.
+def dice_simulation_strings(categories, num_dice, num_rolls, fixed_mult, step_mult, diff):
+    tup = tuple(
+        [
+            tuple(sorted([c.name + str(c.quantity) for c in categories])),
```

Comment:
Yeah sure I'll, try, just know that I will have less time to tell you all about the wondrous world of probability distributions! I was just about to introduce the probability generating function.......

Anyways I have tried what you mentioned but it is x2 slower.
Can you check if I did what you meant? Here is the commit (I reverted it for now):
https://github.com/ArchipelagoMW/Archipelago/pull/3482/commits/34f2c1aed8c8813b2d9c58896650b82a810d3578

Reply:
Awwww, okay. Yeah, if it's slower, then it's not worth it obv :)

Reply:
Oh, the reason I was saying it is because I thought maybe then you could just completely get rid of

```
tup = (
     categories,
     num_dice,
     num_rolls,
     fixed_mult,
     ...
)
```

because now that's just the output of `extract_progression` directly. Like essentially, maybe at this point you could just

```
tup = extract_progression(...)
```

But I don't imagine this will speed things up by a bunch suddenly, so

Reply:
A bit further we sort the tuple/list by meanscore. It seems that sorting a tuple is slower than sorting a list, maybe?

Reply:
Oh, maybe. Well, I tried! Feel free to go without this then :)